packages/@aws-cdk/aws-iam/lib/optimize-statements.ts

```diff
@@ -0,0 +1,286 @@
+/**
+ * Optimizes a list of objects containing IAM policy statements in order to reduce the total size of the rendered JSON
+ * object. The following optimizations are performed:
+ * - Canonicalization of globs in (Not)Resource & (Not)Action: removal entries that are matched by another entry's glob
+ * - De-duplication: removal of statements that have no SID, and are subsets of another statement
+ * - Merging: merging of statements that have no SID, and differ only in either ``Resource``
+ *
+ * Additionally, normalization is performed:
+ * - (Not)Resource & (Not)Action are sorted in alphanumerical order (of the JSON object, if the values are not String)
+ *
+ * @param statements the list of rendered statements to optimize. Those may contain CloudFormation intrinsics.
+ *
+ * @returns a list of equivalents (and hopefylly smaller) statements.
+ */
+export function optimizeStatements(statements: any[]): any[] {
+  const parsed = statements.map(statement => new Statement(statement));
+  return _deduplicate(parsed).map(statement => statement.toJson());
+}
+
+/**
+ * A Policy statement, with convenience functions make it easier to de-duplicate, normalize, ...
+ */
+class Statement implements Deduplicable, Mergeable {
+  private readonly sid?: string;
+
+  private readonly effect: 'Allow' | 'Deny';
+
+  private readonly actions: Glob[];
+  private readonly resources: Glob[];
+
+  private readonly principal: unknown;
+
+  private readonly condition: { [operator: string]: unknown };
+
+  constructor(statement: any) {
+    this.sid = statement.Sid;
+
+    if (statement.Effect !== 'Allow' && statement.Effect !== 'Deny') {
+      throw new Error(`Illegal policy statement effect: ${JSON.stringify(statement.Effect)}`);
+    }
+    this.effect = statement.Effect;
+
+    this.actions = _makeGlobs(statement.Action);
+    this.resources = _makeGlobs(statement.Resource);
+
+    this.principal = statement.Principal || {};
+
+    this.condition = statement.Condition || {};
+
+    function _makeGlobs(expressions: unknown): Glob[] {
+      if (!expressions) { return []; }
+      if (!Array.isArray(expressions)) { return _makeGlobs([expressions]); }
+      return expressions.map(expression => new Glob(expression));
+    }
+  }
+
+  public mergeable(other: Statement, allowSid = false): boolean {
+    return (allowSid || this.sid == null)
+      && this.sid === other.sid
+      && this.effect === other.effect
+      && this._allSuperceded(other.actions, this.actions)
+      && _deepEqual(this.condition, other.condition)
+      && _deepEqual(this.principal, other.principal);
+  }
+
+  public supercedes(other: Statement): boolean {
```

Comment:
Why do we need to make the distinction? `merge` itself could do this:

```
{ s3:GetObject, arn:aws:s3::::* } MERGE { s3:GetObject, arn:aws:s3::::hello } == { s3:GetObject, arn:aws:s3::::* }
```

Then there's no need for `supercedes` anymore.  I imagine (part of?) the reason you have `supercede` today to avoid a scenario like the following:

```
{ s3:GetObject, R } MERGE { s3:GetObject, R } == { s3:GetObject, [R, R]  }
```

So now you have `supercedes` to detect that so you can drop one. But if you gave `merge` more smarts it could do the deduplication there.

Also, I think the word to use here is "subsume".

Reply:
You're right... In fact, the `[R, R]` case is a non-issue because de-duplication happens at the time the document is rendered to JSON anyway.